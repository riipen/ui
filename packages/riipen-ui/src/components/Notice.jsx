import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

class Notice extends React.Component {
  static propTypes = {
    /**
     * The content of the Notice.
     */
    children: PropTypes.node,

    /**
     * List of additional classes to apply to this component.
     */
    classes: PropTypes.array,

    /**
     * The color to use.
     */
    color: PropTypes.oneOf([
      "negative",
      "positive",
      "primary",
      "secondary",
      "tertiary"
    ])
  };

  static defaultProps = {
    classes: [],
    color: "primary"
  };

  static contextType = ThemeContext;

  render() {
    const { classes, children, color } = this.props;

    const theme = this.context;

    const className = clsx(color, classes);

    return (
      <React.Fragment>
        <div className={className}>{children}</div>
        <style jsx>{`
          div {
            border: 1px solid transparent;
            border-radius: ${theme.shape.borderRadius.md};
            opacity: 0.9;
            padding: ${theme.spacing(2)}px ${theme.spacing(4)}px;
          }

          .primary {
            background-color: ${theme.palette.primary.light};
            border-color: ${theme.palette.primary.main};
          }
          .secondary {
            background-color: ${theme.palette.secondary.light};
            border-color: ${theme.palette.secondary.main};
          }
          .tertiary {
            background-color: ${theme.palette.tertiary.light};
            border-color: ${theme.palette.tertiary.main};
          }
          .negative {
            background-color: ${theme.palette.negative.light};
            border-color: ${theme.palette.negative.main};
          }
          .positive {
            background-color: ${theme.palette.positive.light};
            border-color: ${theme.palette.positive.main};
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(Notice);
