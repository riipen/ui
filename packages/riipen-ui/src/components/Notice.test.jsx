import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Notice from "./Notice.jsx";

describe("<Notice>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Notice />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Notice />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new Notice().type.defaultProps;

      const wrapper = mount(<Notice />);

      const component = wrapper.find("Notice");
      expect(component.props().color).toEqual(defaultProps.color);
    });
  });

  describe("children prop", () => {
    it("renders given children", () => {
      const child = <h5>hi hi</h5>;

      const wrapper = mount(<Notice>{child}</Notice>);

      expect(
        wrapper
          .find("Notice")
          .childAt(0)
          .contains(child)
      ).toEqual(true);
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the div inside of Notice", () => {
      const classes = ["classOne"];

      const wrapper = mount(<Notice classes={classes} />);

      expect(
        wrapper
          .find("Notice")
          .find("div")
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-notice"];

      const wrapper = mount(<Notice />);

      expect(
        wrapper
          .find("Notice")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Notice classes={classes} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("color prop", () => {
    it("sets color class name with a valid color", () => {
      const color = "positive";

      const wrapper = mount(<Notice color={color} />);

      expect(
        wrapper
          .find("Notice")
          .childAt(0)
          .hasClass(color)
      ).toEqual(true);
    });

    it("gives an error when given an invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Notice color="red" />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });
});
