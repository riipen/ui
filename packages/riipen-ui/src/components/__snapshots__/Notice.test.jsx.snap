// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notice> renders correct snapshot 1`] = `
<withClasses(Notice)>
  <Notice
    classes={
      Array [
        "riipen",
        "riipen-notice",
      ]
    }
    color="primary"
  >
    <div
      className="jsx-3616308683 primary riipen riipen-notice"
    />
    <JSXStyle
      dynamic={
        Array [
          "4px",
          10,
          20,
          "#2c8ddc",
          "#1f71ae",
          "#f5bf95",
          "#f58632",
          "#78becc",
          "#3caabb",
          "#ea7e7e",
          "#e95353",
          "#78dbac",
          "#4cc88d",
        ]
      }
      id="483255525"
    />
  </Notice>
</withClasses(Notice)>
`;
